quinn-proto/src/cid_queue.rs

```diff
@@ -31,60 +31,83 @@ impl CidQueue {
         }
     }
 
-    pub fn insert(&mut self, cid: IssuedCid) -> Result<(), InsertError> {
-        if cid.sequence == self.offset && self.buffer[self.cursor].is_some() {
-            // This is a duplicate of the active CID.
-            return Ok(());
-        }
-        let index = cid
-            .sequence
-            .checked_sub(self.offset)
-            .ok_or(InsertError::Retired)?;
-        if index >= Self::LEN as u64 {
+    /// Handle a `NEW_CONNECTION_ID` frame
+    ///
+    /// Returns a non-empty range of retired sequence numbers and the reset token of the new active
+    /// CID iff any CIDs were retired.
+    pub fn insert(
+        &mut self,
+        cid: NewConnectionId,
+    ) -> Result<Option<(Range<u64>, ResetToken)>, InsertError> {
+        // Position of new CID wrt. the current active CID
+        let index = match cid.sequence.checked_sub(self.offset) {
+            None => return Err(InsertError::Retired),
+            Some(x) => x,
+        };
+
+        let retired_count = cid.retire_prior_to.saturating_sub(self.offset);
+        if index >= Self::LEN as u64 + retired_count {
             return Err(InsertError::ExceedsLimit);
         }
-        let index = (self.cursor + index as usize) % Self::LEN;
-        self.buffer[index] = Some((cid.id, Some(cid.reset_token)));
-        Ok(())
-    }
 
-    /// Returns the possibly-empty range of newly retired CIDs
-    pub fn retire_prior_to(&mut self, sequence: u64) -> Range<u64> {
-        let n = match sequence.checked_sub(self.offset) {
-            None => return 0..0,
-            Some(n) => n as usize,
-        };
-        for i in 0..n {
+        // Discard retired CIDs, if any
+        for i in 0..(retired_count.min(Self::LEN as u64) as usize) {
             self.buffer[(self.cursor + i) % Self::LEN] = None;
         }
+
+        // Record the new CID
+        let index = ((self.cursor as u64 + index) % Self::LEN as u64) as usize;
+        self.buffer[index] = Some((cid.id, Some(cid.reset_token)));
+
+        if retired_count == 0 {
+            return Ok(None);
+        }
+
+        // The active CID was retired. Find the first known CID with sequence number of at least
```

Comment:
why was it the active one, instead of a lower one? Is the active one always the lowest in the list?

Reply:
Yep; the set of known CIDs is a sliding window into the sequence number space, where the first element in the window is the CID we use.